django/contrib/auth/decorators.py

```diff
@@ -21,19 +40,29 @@ def decorator(view_func):
         def _wrapper_view(request, *args, **kwargs):
             if test_func(request.user):
                 return view_func(request, *args, **kwargs)
-            path = request.build_absolute_uri()
-            resolved_login_url = resolve_url(login_url or settings.LOGIN_URL)
-            # If the login url is the same scheme and net location then just
-            # use the path as the "next" url.
-            login_scheme, login_netloc = urlparse(resolved_login_url)[:2]
-            current_scheme, current_netloc = urlparse(path)[:2]
-            if (not login_scheme or login_scheme == current_scheme) and (
-                not login_netloc or login_netloc == current_netloc
-            ):
-                path = request.get_full_path()
-            from django.contrib.auth.views import redirect_to_login
-
-            return redirect_to_login(path, resolved_login_url, redirect_field_name)
+            return _redirect_to_login(request, login_url, redirect_field_name)
+
+        return _wrapper_view
+
+    return decorator
+
+
+def auser_passes_test(
```

Comment:
Hi, thank you for your comment. Here is my view.

I think what we need is an asynchronous version instead of adapting current functions to support coroutine functions. Here's why: 

1. In Django, IO-related functions usually have distinct coroutine versions, like `aget` and `acreate`. 

2. Having a dedicated async decorator allows for more controlled, simpler, and user-friendly behavior. If we modify existing sync function decorators to handle coroutine functions, it raises a question: should they return a coroutine or a sync function when decorating a sync function? My view is that they should return a sync function in a WSGI environment and a coroutine function in an ASGI environment. This approach, however, complicates the decorator's logic and makes it less intuitive. On the other hand, using a new coroutine function decorator keeps things straightforward. For instance, a sync function decorator would always return a sync function, and a coroutine function decorator would always return a coroutine function.

Do you think these are good reasons? Welcome to discuss.

Reply:
There is a huge difference, we don't have async decorators, we make them to handle async functions passed as parameters. So creating a separate functions with the `a` prefix is confusing and inconsistent with other decorators.

Reply:
OK, I understand what you mean.  Is there any similar case in Django?  Could you please give me an example for reference?

Reply:
You can check ticket-31949 and all changed decorators.

Reply:
> You can check [ticket-31949](https://code.djangoproject.com/ticket/31949) and all changed decorators.

Okay, I'll take a look at this ticket, thank you.

Reply:
Hi, I have read ticket-31949, I have understand what you mean. But I still have a question. For example, there is a very simple sync view, it returns HttpResponse("OK") directly. But now I decide to decorate it by using the `login_required` decorator. According to the ticket-31949, if the view is a function, the decorator will return a function too. In the end, we got a function that includes database I/O, which will run in the async context. I think it's really strange.

```
# Assuming this is an async backend.
@login_required
def foo(request):
    return HttpResponse(”OK“)
```

Reply:
Hi, I have read ticket-31949, I have understand what you mean. But I still have a question, becauce I noticed that most of the decorators referenced in the ticket-31949 do not contain any I/O, but the `login_required` contains I/O. So I think there are some different. For example, there is a very simple sync view, it returns `HttpResponse("OK")` directly. But now I decide to decorate it by using the `login_required` decorator. According to the ticket-31949, if the view is a function, the decorator will return a function too. In the end, we got a function that includes synchronous database I/O, which will run in the async context. I think it's really strange.

```
# Assuming this is an async backend.
# The decorated foo is a function that will query the database synchronously.

@login_required
def foo(request):
    return HttpResponse(”OK“)
```

Although Django has technology that allows synchronous view functions to run in async context, I still feel that this is not as good as native coroutine functions.

Reply:
I'm not sure what are you trying to say, TBH :thinking: Modified `@login_required` should recognized if decorated function is coroutine or not, and take different paths. This not much different from what you proposed, the main difference is that we keep one decorator that can handle both.

Reply:
Hi, could you please view my simple example that explains my view? I have created a repo here https://github.com/HappyDingning/dango_playground/tree/main

I implemented a demo `login_required`, that can decorate the view function according to its type(coroutine or not). And I also implemented a demo `alogin_required`, that can fix the issue of `login_required`.
Please read the comments in the [views.py ](https://github.com/HappyDingning/dango_playground/blob/a181bcd1e7f7798910e1e1ccac873acf1151727a/playground/views.py#L9)and [decorators.py(line 62~63)](https://github.com/HappyDingning/dango_playground/blob/a181bcd1e7f7798910e1e1ccac873acf1151727a/playground/decorators.py#L62). Thank you~

Reply:
I really don't understand what you're trying to achieve you've created async versions of decorators and trying to decorated sync function views with them :exploding_head: I'm puzzled.

Reply:
This ticket is about modifying `login_required` and `permission_required` and `user_passes_test` that already handle async functions, to be able to handle coroutines, that's it.

> Please read the comments in the [views.py ](https://github.com/HappyDingning/dango_playground/blob/89f8bd4b1bf1b61022174c015c4caedb75369bcb/playground/views.py#L9)and [decorators.py(line 62~63)](https://github.com/HappyDingning/dango_playground/blob/a181bcd1e7f7798910e1e1ccac873acf1151727a/playground/decorators.py#L62). Thank you~

I think you're looking on this wrong the wrong perspective `test_func` is not important, `view_func` is decorated and we want to handle when it's a coroutine:

```python
def user_passes_test(
    test_func, login_url=None, redirect_field_name=REDIRECT_FIELD_NAME
):
    """
    Decorator for views that checks that the user passes the given test,
    redirecting to the log-in page if necessary. The test should be a callable
    that takes the user object and returns True if the user passes.
    """

    def decorator(view_func):
        if iscoroutinefunction(view_func):
            
            async def _view_wrapper(request, *args, **kwargs):
                ...
             
        else:
            
            def _view_wrapper(request, *args, **kwargs):
                ...
                
        return wraps(view_func)(_view_wrapper)
        
    return decorator

```

Reply:
I'm really sorry for confusing you. I created a post. You are welcome to leave a message. https://forum.djangoproject.com/t/about-async-version-of-the-login-required-decorator/27077